src/scipp/core/data_group.py

```diff
@@ -23,43 +25,106 @@ def _summarize(item):
+    A dict-like group of data. Additionally provides dims and shape properties.
+
+    DataGroup acts like a Python dict but additionally supports Scipp functionality
+    such as positional- and label-based indexing and Scipp operations by mapping them
+    to the values in the dict. This may happen recursively to support tree-like data
+    structures.
     """
 
-    def __init__(self, items=None):
+    def __init__(self, items=None, /):
         self._items = dict(items) if items is not None else {}
 
     def __len__(self) -> int:
+        """Return the number of items in the data group."""
         return len(self._items)
 
     def __iter__(self):
         yield from self._items
 
+    @overload
+    def __getitem__(self, name: str) -> Any:
+        ...
+
+    @overload
+    def __getitem__(self, name: ScippIndex) -> DataGroup:
+        ...
+
     def __getitem__(self, name):
+        """Return item of given name or index all items.
+
+        When ``name`` is a string, return the items of the given name. Otherwise, this
+        returns a new DataGroup, with items created by indexing the items in this
+        DataGroup. This may perform, e.g., Scipp's positional indexing or label-based
+        indexing on items that are scipp.Variable or scipp.DataArray.
+
+        Positional indexing is only possible when the shape of all items is consistent
+        for the indexed dimension.
+
+        Label-based indexing is only possible when all items have a coordinate for the
+        indexed dimension.
+        """
         if isinstance(name, str):
             return self._items[name]
-        if name == ():
+        if isinstance(name, tuple) and name == ():
             return DataGroup({key: var[()] for key, var in self.items()})
-        dim, index = name
-        if isinstance(index, int) and self.sizes[dim] is None:
-            raise ValueError(
-                f"Positional indexing dim {dim} not possible as the length is not "
+        if _is_positional_index(name):
+            if self.ndim != 1:
+                raise DimensionError(
+                    "Slicing with implicit dimension label is only possible "
+                    f"for 1-D objects. Got {self.sizes} with ndim={self.ndim}. Provide "
+                    "an explicit dimension label, e.g., var['x', 0] instead of var[0].")
+            dim = self.dims[0]
+            index = name
+        else:
+            dim, index = name
+        if _is_positional_index(index) and self.sizes[dim] is None:
+            raise DimensionError(
+                f"Positional indexing dim '{dim}' not possible as the length is not "
                 "unique.")
         return DataGroup({
             key: var[dim, index] if dim in _item_dims(var) else var
             for key, var in self.items()
         })
 
+    @overload
+    def __setitem__(self, name: str, value: Any):
+        ...
+
+    @overload
+    def __setitem__(self, name: ScippIndex, value: DataGroup):
```

Comment:
This isn't implemented?

Reply:
Indeed, should I remove the overload for now?

Reply:
yes